x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ApiKeyService.java

```diff
@@ -1083,47 +1192,42 @@ private void doUpdateApiKey(
-                request.getId(),
+                currentVersionedDoc.id(),
                 currentDocVersion,
                 targetDocVersion
             );
         }
-
         final XContentBuilder builder = maybeBuildUpdatedDocument(
+            currentVersionedDoc.id(),
             currentVersionedDoc.doc(),
             targetDocVersion,
             authentication,
             request,
             userRoleDescriptors
         );
         final boolean isNoop = builder == null;
-        if (isNoop) {
-            logger.debug("Detected noop update request for API key [{}]. Skipping index request.", request.getId());
-            listener.onResponse(new UpdateApiKeyResponse(false));
-            return;
-        }
+        return isNoop
+            ? null
+            : client.prepareIndex(SECURITY_MAIN_ALIAS)
+                .setId(currentVersionedDoc.id())
+                .setSource(builder)
+                .setIfSeqNo(currentVersionedDoc.seqNo())
+                .setIfPrimaryTerm(currentVersionedDoc.primaryTerm())
+                .setOpType(DocWriteRequest.OpType.INDEX)
+                .request();
+    }
 
-        final IndexRequest indexRequest = client.prepareIndex(SECURITY_MAIN_ALIAS)
-            .setId(request.getId())
-            .setSource(builder)
-            .setIfSeqNo(currentVersionedDoc.seqNo())
-            .setIfPrimaryTerm(currentVersionedDoc.primaryTerm())
-            .setOpType(DocWriteRequest.OpType.INDEX)
-            .request();
-        logger.trace("Executing index request to update API key [{}]", request.getId());
-        securityIndex.prepareIndexIfNeededThenExecute(
-            ex -> listener.onFailure(traceLog("prepare security index before update", ex)),
-            () -> executeAsyncWithOrigin(
-                client.threadPool().getThreadContext(),
-                SECURITY_ORIGIN,
-                client.prepareBulk().add(indexRequest).setRefreshPolicy(RefreshPolicy.WAIT_UNTIL).request(),
-                ActionListener.<BulkResponse>wrap(
-                    bulkResponse -> translateResponseAndClearCache(request.getId(), bulkResponse, listener),
-                    ex -> listener.onFailure(traceLog("execute bulk request for update", ex))
-                ),
-                client::bulk
-            )
-        );
+    private void addErrorsForNotFoundApiKeys(
+        final BulkUpdateApiKeyResponse.Builder responseBuilder,
+        final Collection<VersionedApiKeyDoc> foundDocs,
+        final List<String> requestedIds
+    ) {
+        final Set<String> foundIds = foundDocs.stream().map(VersionedApiKeyDoc::id).collect(Collectors.toUnmodifiableSet());
+        for (String id : requestedIds) {
+            if (foundIds.contains(id) == false) {
+                responseBuilder.error(id, new ResourceNotFoundException("no API key owned by requesting user found for ID [" + id + "]"));
+            }
+        }
```

Comment:
I think we can compare size first and short circuit if size is the same.

Reply:
Good call! I considered this originally, then got into over-thinking this when it comes to duplicates, then forgot about it 😅  Will add the short-circuit check.